tsdb/querier.go

```diff
@@ -793,23 +800,38 @@ func (p *populateWithDelChunkSeriesIterator) Next() bool {
 			t, h = p.currDelIter.AtHistogram()
 
 			// Defend against corrupted chunks.
-			pI, nI, okToAppend, counterReset := app.(*chunkenc.HistogramAppender).Appendable(h)
-			if len(pI)+len(nI) > 0 {
-				err = fmt.Errorf(
-					"bucket layout has changed unexpectedly: %d positive and %d negative bucket interjections required",
-					len(pI), len(nI),
-				)
-				break
-			}
-			if counterReset {
-				err = errors.New("detected unexpected counter reset in histogram")
-				break
-			}
-			if !okToAppend {
-				err = errors.New("unable to append histogram due to unexpected schema change")
-				break
+			switch h.CounterResetHint {
+			case histogram.GaugeType:
+				pI, nI, bpI, bnI, _, _, okToAppend := app.(*chunkenc.HistogramAppender).AppendableGauge(h)
+				if len(pI)+len(nI)+len(bpI)+len(bnI) > 0 {
+					err = fmt.Errorf(
+						"bucket layout has changed unexpectedly: forward %d positive, %d negative, backward %d positive %d negative bucket interjections required",
+						len(pI), len(nI), len(bpI), len(bnI),
+					)
+					break loopH
+				}
+				if !okToAppend {
+					err = errors.New("unable to append histogram due to unexpected schema change")
+					break loopH
+				}
+			default:
+				pI, nI, okToAppend, counterReset := app.(*chunkenc.HistogramAppender).Appendable(h)
+				if len(pI)+len(nI) > 0 {
+					err = fmt.Errorf(
+						"bucket layout has changed unexpectedly: %d positive and %d negative bucket interjections required",
+						len(pI), len(nI),
+					)
+					break loopH
+				}
+				if counterReset {
+					err = errors.New("detected unexpected counter reset in histogram")
```

Comment:
Yes, feels a bit arbitrary to trust the sample over the header. On the other hand this "fix" only works if all samples are gauge histograms, since otherwise Appendable() call will fail due to this [condition](https://github.com/prometheus/prometheus/blob/7c2de14b0bd74303c2ca6f932b71d4585a29ca75/tsdb/chunkenc/histogram.go#L267).

So it's potentially 100 samples versus one header. Which to me begs the question why do we take into account the header in [AppendableGauge](https://github.com/prometheus/prometheus/blob/7c2de14b0bd74303c2ca6f932b71d4585a29ca75/tsdb/chunkenc/histogram.go#L339). We could just look at the previous sample and check if it's a gauge.
Going further, do we even have to set the reset hint to other than "unknown" for gauge histograms?

Reply:
I don't think we store the "gauge-ness" of a histogram in the samples within the chunks. The only place where we store the information that the histograms in the chunks are gauge histograms _is_ the counter reset header.

An "unknown" counter reset header means "the histograms in this chunk are definitely counter histograms but it is unknown if the chunk was started because of a counter reset"